plugins/c8y_firmware_plugin/src/tests.rs

```diff
@@ -0,0 +1,775 @@
+    mqtt_tests::assert_received_all_expected(
+        &mut smartrest_messages,
+        TEST_TIMEOUT_MS,
+        &[
+            "501,c8y_Firmware",
+            format!("115,{FIRMWARE_NAME},{FIRMWARE_VERSION},{cloud_firmware_url}").as_str(),
+            "503,c8y_Firmware",
+        ],
+    )
+    .await;
+
+    Ok(())
+}
+
+#[tokio::test(flavor = "multi_thread", worker_threads = 4)]
+#[serial_test::serial]
+async fn handle_response_executing_and_failed_child_device() -> anyhow::Result<()> {
+    let mut tmp_dir = TempTedgeDir::new();
+    create_required_directories(&mut tmp_dir);
+
+    let broker = mqtt_tests::test_mqtt_broker();
+    let mut c8y_http_client = MockC8YHttpProxy::new();
+    c8y_http_client
+        .expect_download_file()
+        .with(
+            predicate::always(),
+            predicate::always(),
+            predicate::always(),
+        )
+        .returning(|_, file_name, tmp_dir_path| {
+            let downloaded_path = tmp_dir_path.join(file_name);
+            std::fs::File::create(&downloaded_path)?;
+            Ok(downloaded_path)
+        });
+
+    start_firmware_manager(
+        &mut tmp_dir,
+        broker.port,
+        c8y_http_client,
+        DEFAULT_REQUEST_TIMEOUT_SEC,
+    )
+    .await?;
+
+    // Mock download endpoint for the plugin to download a firmware from the cloud
+    let mock_http_server_host = mockito::server_url();
+    let cloud_firmware_url = format!("{mock_http_server_host}/some/cloud/url");
+
+    // Subscribe tedge request endpoint for firmware update
+    let mut tedge_command_messages = broker
+        .messages_published_on(&format!(
+            "tedge/{CHILD_DEVICE_ID}/commands/req/firmware_update"
+        ))
+        .await;
+
+    // Publish a c8y_Firmware operation to the plugin.
+    broker
+        .publish(
+            "c8y/s/ds",
+            format!(
+                "515,{CHILD_DEVICE_ID},{FIRMWARE_NAME},{FIRMWARE_VERSION},{cloud_firmware_url}"
+            )
+            .as_str(),
+        )
+        .await?;
+
+    // Firmware update REQUEST should be received.
+    let received_message = tedge_command_messages
+        .next()
+        .with_timeout(TEST_TIMEOUT_MS)
+        .await?
+        .expect("No message received.");
+    let received_json: serde_json::Value = serde_json::from_str(&received_message)?;
+    let operation_id = received_json.get("id").unwrap().as_str().unwrap();
+
+    // Subscribe SmartREST endpoint for child device
+    let mut smartrest_messages = broker
+        .messages_published_on(format!("c8y/s/us/{CHILD_DEVICE_ID}").as_str())
+        .await;
+
+    // Publish a executing RESPONSE from child device
+    broker
+        .publish(
+            &format!("tedge/{CHILD_DEVICE_ID}/commands/res/firmware_update"),
+            &json!({
+                "status": "executing",
+                "id": operation_id,
+            })
+            .to_string(),
+        )
+        .await?;
+
```

Comment:
I'd assert the `501` message here to make sure that it is sent immediately in response to the `executing` response from the child.